text/0000-default-field-values.md

```diff
@@ -0,0 +1,1977 @@
+    pub ipsum: u32,
+}
+```
+
+### Conclusion
+
+As seen in the previous sections, rather than make deriving `Default`
+more magical, by allowing default field values in the language,
+user-space custom derive macros can make use of them.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Providing field defaults
+
+Consider a data-type such as (1):
+
+```rust
+pub struct Probability {
+    value: f32,
+}
+```
+
+You'd like encode the default probability value to be `0.5`;
+With this RFC now you can provide such a default directly where `Probability`
+is defined like so (2):
+
+```rust
+pub struct Probability {
+    value: f32 = 0.5,
+}
+```
+
+Having done this, you can now construct a `Probability` with a struct
+initializer and leave `value` out to use the default (3):
+
+```rust
+let prob = Probability { .. };
+```
+
+## Deriving `Default`
+
+Previously, you might have instead implemented the `Default` trait like so (4):
+
+```rust
+impl Default for Probability {
+    fn default() -> Self {
+        Self { value: 0.5 }
+    }
+}
+```
+
+You can now shorten this to (5):
+
+```rust
+impl Default for Probability {
+    fn default() -> Self {
+        Self { .. }
+    }
+}
+```
+
+However, since you had specified `value: f32 = 0.5` in the definition of
+`Probability`, you can take advantage of that to write the more simpler
+and more idiomatic (6):
+
+```rust
+#[derive(Default)]
+pub struct Probability {
+    value: f32 = 0.5,
+}
+```
+
+Having done this, a `Default` implementation equivalent to the one in (5)
+will be generated for you.
+
+## More fields
+
+As you saw in the [summary], you are not limited to a single field and all
+fields need not have any defaults associated with them. Instead, you can freely
+mix and match. Given the definition of `LaunchCommand` from the [motivation] (7):
+
+```rust
+struct LaunchCommand {
+    cmd: String,
+    args: Vec<String> = Vec::new(),
+    some_special_setting: Option<FancyConfig> = None,
+    setting_most_people_will_ignore: Option<FlyMeToTheMoon> = None,
+}
+```
+
+you can omit all fields but `cmd` (8):
+
+```rust
+let ls_cmd = LaunchCommand {
+    cmd: "ls".to_string(),
+    ..
+};
+```
+
+You can also elect to override the provided defaults (9):
+
+```rust
+let ls_cmd2 = LaunchCommand {
+    cmd: "ls".to_string(),
+    args: vec!["-lah".to_string()],
+    some_special_setting: make_special_setting(),
+    // setting_most_people_will_ignore is still defaulted.
+    ..
+};
+```
+
+## Default fields values are [`const` context]s
+
+[`const` context]: https://github.com/rust-lang-nursery/reference/blob/66ef5396eccca909536b91cad853f727789c8ebe/src/const_eval.md#const-context
+
+As you saw in (7), `Vec::new()`, a function call, was used.
+However, this assumes that `Vec::new` is a *`const fn`*. That is, when you
+provide a default value `field: Type = value`, the given `value` must be a
+*constant expression* such that it is valid in a [`const` context].
```

Comment:
What are the other requirements on the value type? `Vec` example indicates the type doesn't need to be `Copy`. Does it need to be `Clone`? The decision here might be especially important for ZSTs which are used as some kind of token and are supposed to guarantee that they cannot be conjured/cloned out of thin air

Reply:
> What are the other requirements on the value type? `Vec` example indicates the type doesn't need to be `Copy`. Does it need to be `Clone`?

I don't think there should be. The value is given as a const expression, which can be thought as being constructed on each usage. This is also why `[CONSTANT; 10]` always work and does not require the element type to be `Clone` or `Copy`.

> The decision here might be especially important for ZSTs which are used as some kind of token and are supposed to guarantee that they cannot be conjured/cloned out of thin air

If the definition site explicitly mark a default field by using a valid way to construct it. It does not create any extra way to conjure a ZST. It's the same as they define a `impl S { pub const DEFAULT_FIELD: ArbitraryNoCopyTy = some_constructor(); }` and users write `S { field: S::DEFAULT_FIELD }`.